### PR TITLE
refactor CUDA gencode flags, add "common archs" option

### PR DESCRIPTION
**Pull request by yxlao**
This PR servers 3 purposes:
1. Simplify CUDA architecture flags, with CMake 3.18+
2. Support `stdgpu`
3. Move `open3d_show_and_abort_on_warning` to the root `CMakeLists.txt`

This change is [Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3228)

### CODE REVIEW

**Comment by yxlao** [comment]
New issue poped up: thrust not found for STDGPU.
Will remove draft state once it is ready.

**Comment by yxlao** [comment]
according to experiments in #3230 , we need to install the full `cuda-10-1` (or other versions) package to enable thrust,

**Review by benjaminum — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3228)* status: 0 of 5 files reviewed, 3 unresolved discussions (waiting on @theNded and @yxlao)

___
*[CMakeLists.txt, line 21 at r3](https://reviewable.io/reviews/intel-isl/open3d/3228#-MX1gjDeBUYUcMFMWuPa:-MX1gjDeBUYUcMFMWuPb:b-fbd7k2) ([raw file](https://github.com/intel-isl/open3d/blob/71d9d943df3e4e7b95b026ea2040bf0f2c784650/CMakeLists.txt#L21)):*
> ```CMake
> option(BUILD_CUDA_MODULE          "Build the CUDA module"                    OFF)
> option(BUILD_CACHED_CUDA_MANAGER  "Build the cached CUDA memory manager"     ON )
> option(BUILD_COMMON_CUDA_ARCHS    "Build for common CUDA GPUs (for relase)"  OFF)
> ```

typo release.
___
*[CMakeLists.txt, line 338 at r3](https://reviewable.io/reviews/intel-isl/open3d/3228#-MX1gQCSClDIUl2kwxua:-MX1gQCSClDIUl2kwxub:bs8ixax) ([raw file](https://github.com/intel-isl/open3d/blob/71d9d943df3e4e7b95b026ea2040bf0f2c784650/CMakeLists.txt#L338)):*
> ```CMake
>     enable_language(CUDA)
>     if (CMAKE_CUDA_COMPILER)
>         find_package(CUDA REQUIRED)
> ```

Is this still needed? It is deprecated and I remember the only reason for it was to detect the supported CUDA_ARCH.
___
*[CMakeLists.txt, line 366 at r3](https://reviewable.io/reviews/intel-isl/open3d/3228#-MX1k7ff0tpFULsodZHO:-MX1k7ff0tpFULsodZHP:b-umcx16) ([raw file](https://github.com/intel-isl/open3d/blob/71d9d943df3e4e7b95b026ea2040bf0f2c784650/CMakeLists.txt#L366)):*
> ```CMake
>     if (BUILD_COMMON_CUDA_ARCHS)
>         if (CMAKE_CUDA_COMPILER_VERSION VERSION_GREATER_EQUAL "11.0")
>             set(CMAKE_CUDA_ARCHITECTURES 35 50 52 60 61 70 75 80)
> ```

According to the cmake documentation this will include ptx code for all archs. I think we want to include ptx only for the latest arch to be compatible with future archs/drivers.
We can check with ```cuobjdump```.

**Review by yxlao — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3228)* status: 0 of 5 files reviewed, 3 unresolved discussions (waiting on @benjaminum and @theNded)

___
*[CMakeLists.txt, line 21 at r3](https://reviewable.io/reviews/intel-isl/open3d/3228#-MX1gjDeBUYUcMFMWuPa:-MX2oIXYB1mLgRO56iis:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/71d9d943df3e4e7b95b026ea2040bf0f2c784650/CMakeLists.txt#L21)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

typo release.
</blockquote></details>

Done.
___
*[CMakeLists.txt, line 338 at r3](https://reviewable.io/reviews/intel-isl/open3d/3228#-MX1gQCSClDIUl2kwxua:-MX2oJ1EAoMMxILzfrrT:bxo4hyt) ([raw file](https://github.com/intel-isl/open3d/blob/71d9d943df3e4e7b95b026ea2040bf0f2c784650/CMakeLists.txt#L338)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

Is this still needed? It is deprecated and I remember the only reason for it was to detect the supported CUDA_ARCH.
</blockquote></details>

This is required by NPP in `find_dependencies.cmake`. I put it here in case other dependencies in the future would also require this. Added a comment in code.
___
*[CMakeLists.txt, line 366 at r3](https://reviewable.io/reviews/intel-isl/open3d/3228#-MX1k7ff0tpFULsodZHO:-MX2oXgw7a90tzO33kQM:bo4zpij) ([raw file](https://github.com/intel-isl/open3d/blob/71d9d943df3e4e7b95b026ea2040bf0f2c784650/CMakeLists.txt#L366)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

According to the cmake documentation this will include ptx code for all archs. I think we want to include ptx only for the latest arch to be compatible with future archs/drivers.
We can check with ```cuobjdump```.
</blockquote></details>

This is true. Updated. 

Now the generated flags for CUDA 11 are:

```
--generate-code=arch=compute_35,code=[sm_35] 
--generate-code=arch=compute_50,code=[sm_50] 
--generate-code=arch=compute_52,code=[sm_52] 
--generate-code=arch=compute_60,code=[sm_60] 
--generate-code=arch=compute_61,code=[sm_61] 
--generate-code=arch=compute_70,code=[sm_70] 
--generate-code=arch=compute_75,code=[sm_75] 
--generate-code=arch=compute_80,code=[compute_80,sm_80]
```

File: CMakeLists.txt [hunk 1]
```diff
@@ -18,6 +18,7 @@ option(BUILD_BENCHMARKS           "Build the micro benchmarks"               OFF
 option(BUILD_PYTHON_MODULE        "Build the python module"                  ON )
 option(BUILD_CUDA_MODULE          "Build the CUDA module"                    OFF)
 option(BUILD_CACHED_CUDA_MANAGER  "Build the cached CUDA memory manager"     ON )
+option(BUILD_COMMON_CUDA_ARCHS    "Build for common CUDA GPUs (for release)" OFF)
```
[the remark this answers is not in the document]
Comment (yxlao) [reply CMakeLists.txt]: Removed `BUILD_COMMON_CUDA_ARCHS`, used `DEVELOPER_BUILD` directly.

**Comment by yxlao** [comment]
New issues detected: although tests are passing in this branch, it turns out that the current mechanism triggers PTX JIT compilation when `stdgpu` functions are called, sometimes these compilation time are prohibitively long. We need a way to synchronize CUDA flags between the main Open3D library and the external library. 

This PR is no longer blocking https://github.com/intel-isl/Open3D/pull/3211. I belive there's a minimal fix to make in https://github.com/intel-isl/Open3D/pull/3211 work already.

**Review by ssheorey — APPROVED** [review]

**Comment by benjaminum** [comment]
Closing because development of this has been finished in https://github.com/intel-isl/Open3D/pull/3316.

**Pull request closed without merging**